### ISSUE TITLE
[Bug]: sb upgrade (prerelease) does not upgrade satellite packages

### ISSUE THREAD

**Issue by yannbf** [storybookjs/storybook#25731]
### Describe the bug

Back in Storybook 7.5.0, there was an indicator of every package and version that would be updated:
![image](https://github.com/storybookjs/storybook/assets/1671563/c9fd9d95-49a6-4d7b-80ad-615a209f2c2c)

Whereas in Storybook 8, there is a banner that just states that Storybook will be upgraded from Y to X version (assuming that all storybook dependencies are using the same version)
<img width="616" alt="image" src="https://github.com/storybookjs/storybook/assets/1671563/a1ed622f-74fb-4562-9e81-55d04dcdab07">

However, apart from most of the `@storybook/*` dependencies which follow the same versioning, there are the so called "satellite packages" which are just storybook packages or addons (official or from the community) that follow a different versioning strategy. In Storybook 8, they are not taken into account when upgrading.

### To Reproduce

Using [MealDrop](https://github.com/yannbf/mealdrop) as a repro, after running `npx storybook@next upgrade`, this is what you get:
<img width="1096" alt="image" src="https://github.com/storybookjs/storybook/assets/1671563/9dcb9c6e-affc-4c03-b1ae-5051d99890a4">

And this is what you were supposed to get (based on 7.5.0 functionality) - notice the test-runner and msw addon being updated to prerelease versions:
<img width="1129" alt="image" src="https://github.com/storybookjs/storybook/assets/1671563/2e1fc5c9-88b2-4293-bbe9-745ee28cba28">

### Additional context

If the satellite packages are not upgraded, especially in the prerelease example, users will get a possibly broken Storybook because they will contain different major versions installed (and will get a warning via the upgrade command):

<img width="443" alt="image" src="https://github.com/storybookjs/storybook/assets/1671563/4432127a-034f-4271-b06e-6a17e65b41a6">

**Comment by shilman** [comment]
This is the decision we made in https://github.com/storybookjs/storybook/issues/25293 cc @JReinhold 

How does fixing this for satellite repos help? How are they different from third party addons in the grand scheme of things?

We need a more robust solution that solves for all the scenarios. My proposal is to:
1. Run `doctor` automatically after an upgrade
2. Improve `doctor` to (1) not be hard-coded, and (2) provide better diagnostic messages explaining what you need to upgrade

IMHO solving that is the better use of time, and it would subsume this issue. We could still do this one as an optimization, but it would be more of a nice to have than an S2.

**Comment by yannbf** [comment]
> This is the decision we made in #25293 cc @JReinhold
> 
> How does fixing this for satellite repos help? How are they different from third party addons in the grand scheme of things?
> 
> We need a more robust solution that solves for all the scenarios. My proposal is to:
> 
> 1. Run `doctor` automatically after an upgrade
> 2. Improve `doctor` to (1) not be hard-coded, and (2) provide better diagnostic messages explaining what you need to upgrade
> 
> IMHO solving that is the better use of time, and it would subsume this issue. We could still do this one as an optimization, but it would be more of a nice to have than an S2.

Sure that sounds good, sounds like a better idea instead. But ideally we do it before GA, otherwise there will be tons of people with duplicated Storybook dependencies

**Comment by vanessayuenn** [comment]
Copied a proposed solution for step 2 from a notion doc: https://github.com/storybookjs/storybook/issues/26015

@JReinhold can you remind me why it wasn't possible to add running `doctor` after upgrade? I recalled we were going to do that but you found out it wasn't as trivial as expected.

**Comment by vanessayuenn** [comment]
@yannbf isn't this already addressed? Can it be closed?

**Comment by ndelangen** [comment]
My idea for this is the following:

When the user is upgrading to the latest version of storybook;
...then we add an extra automigration

The automigration asks if the user would like us to upgrade storybook related dependencies.

In the question.. we will tell the user that:
- our detection of what is storybook related isn't perfect
- This may not be able to upgrade everything
- We'll run `dedupe` afterwards, and this might cause unintended side-effects

The automigration asks if the user would like us to do this, and by default the answer is "NO".

The user must choose to opt-in to this mechanic (and they only get this option/choice when upgrading to LATEST.

We'll use yann's recent work on detection version conflicts to get a list of packages that have storybook packages as a peerDep or regular dep.

Then when we have that list, 
- we ask the packageManager what the latest version of those packages are
- we ask the packageManager to upgrade those dependencies
- we run `dedupe` afterwards.

@yannbf I assume it will be relatively easy for this future automigration to make use of your recent work for doctor? I'm referring to the getting a list of packages that they themselves depend on storybook? 🙏
